### PR TITLE
perf: Avoid unnecessary data type conversions for DeepSeek-V3 on Blackwell

### PR DESCRIPTION
**Pull request by jinyangyuan-nvidia**
Optimize the performance of DeepSeek-V3 on Blackwell for small batch sizes by

- Change the output data type of router GEMM to FP32
- Convert the data type of `correction_bias` in the object initialization
- ~~Use the `cutlass` backend for the first GEMM in shared experts~~ (only benefits small batches)

The performance has been tested on B200 using the following commands.
```
python3 -m sglang.launch_server \
    --model-path ${path_model} \
    --trust-remote-code \
    --quantization modelopt_fp4 \
    --tp 8 \
    --moe-runner-backend flashinfer_trtllm \
    --attention-backend trtllm_mla \
    --enable-flashinfer-allreduce-fusion

for batch_size in 1 2 4 8 16 32 64 128 256; do
    python3 -m sglang.bench_serving \
        --model ${path_model} \
        --dataset-name random \
        --backend sglang-oai \
        --random-range-ratio 1 \
        --random-input-len 1200 \
        --random-output-len 256 \
        --max-concurrency ${batch_size} \
        --num-prompts $(( batch_size * 10 ))
done
```

The tested median TTFT and median ITL are listed below. Because TTFT has very large variance (comparing it is not very meaningful), the performance is assessed using ITL.

<table>
    <thead>
        <tr>
            <th rowspan=3>Batch Size</th>
            <th rowspan=2, colspan=2>Original</th>
            <th colspan=6>Avoid Dtype Conversion</th>
        </tr>
        <tr>
            <th colspan=2>cuDNN, cuDNN</th>
            <th colspan=2>CUTLASS, cuDNN</th>
            <th colspan=2>CUTLASS, CUTLASS</th>
        </tr>
        <tr>
            <th>TTFT (ms)</th>
            <th>ITL (ms)</th>
            <th>TTFT (ms)</th>
            <th>ITL (ms)</th>
            <th>TTFT (ms)</th>
            <th>ITL (ms)</th>
            <th>TTFT (ms)</th>
            <th>ITL (ms)</th>
        </tr>
    </thead>
    <tbody>
<tr><td align=center>1</td> <td align=center>72.63</td> <td align=center>7.41</td> <td align=center>72.21</td> <td align=center>7.15</td> <td align=center>74.54</td> <th align=center>7.07</th> <td align=center>69.93</td> <td align=center>7.15</td></tr>
<tr><td align=center>2</td> <td align=center>144.49</td> <td align=center>7.41</td> <td align=center>119.35</td> <th align=center>7.12</th> <td align=center>84.18</td> <th align=center>7.12</th> <td align=center>136.82</td> <td align=center>7.14</td></tr>
<tr><td align=center>4</td> <td align=center>159.16</td> <td align=center>7.93</td> <td align=center>156.64</td> <td align=center>7.70</td> <td align=center>173.00</td> <th align=center>7.62</th> <td align=center>151.52</td> <td align=center>7.73</td></tr>
<tr><td align=center>8</td> <td align=center>271.39</td> <td align=center>8.92</td> <td align=center>308.45</td> <th align=center>8.57</th> <td align=center>273.02</td> <td align=center>8.70</td> <td align=center>260.27</td> <td align=center>8.79</td></tr>
<tr><td align=center>16</td> <td align=center>472.31</td> <td align=center>11.70</td> <td align=center>474.21</td> <td align=center>11.41</td> <td align=center>481.33</td> <th align=center>11.23</th> <td align=center>466.03</td> <td align=center>11.33</td></tr>
<tr><td align=center>32</td> <td align=center>808.20</td> <td align=center>14.24</td> <td align=center>561.89</td> <th align=center>14.08</th> <td align=center>559.31</td> <td align=center>14.23</td> <td align=center>544.00</td> <td align=center>14.19</td></tr>
<tr><td align=center>64</td> <td align=center>1515.34</td> <td align=center>18.05</td> <td align=center>1611.65</td> <th align=center>17.89</th> <td align=center>1012.22</td> <td align=center>18.00</td> <td align=center>967.13</td> <td align=center>17.97</td></tr>
<tr><td align=center>128</td> <td align=center>2074.25</td> <td align=center>23.11</td> <td align=center>1890.56</td> <th align=center>22.97</th> <td align=center>1513.91</td> <td align=center>23.05</td> <td align=center>1155.98</td> <td align=center>23.06</td></tr>
<tr><td align=center>256</td> <td align=center>4009.33</td> <td align=center>31.33</td> <td align=center>3878.77</td> <th align=center>31.18</th> <td align=center>3866.18</td> <th align=center>31.18</th> <td align=center>3882.10</td> <td align=center>31.21</td></tr>
    </tbody>
</table>

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by Ying1123** [comment]
Is there any performance data to show the improvement?

File: python/sglang/srt/layers/linear.py [hunk 1]
```diff
@@ -418,7 +421,12 @@ def forward(self, input_):
 
         # Matrix multiply.
         assert self.quant_method is not None
-        output_parallel = self.quant_method.apply(self, input_, bias)
+        if isinstance(self.quant_method, ModelOptFp4LinearMethod):
```
Comment (Ying1123) [inline python/sglang/srt/layers/linear.py]: Avoid using "if...else" here. Maybe add a util function to determine the kwargs of `quant_method.apply`

File: python/sglang/srt/layers/linear.py [hunk 2]
```diff
@@ -1317,7 +1331,17 @@ def forward(self, input_, skip_all_reduce=False):
         # bias will not get added more than once in TP>1 case)
         bias_ = None if (self.tp_rank > 0 or self.skip_bias_add) else self.bias
         with use_symmetric_memory(parallel_state.get_tp_group()) as sm:
-            output_parallel = self.quant_method.apply(self, input_parallel, bias=bias_)
+            if isinstance(self.quant_method, ModelOptFp4LinearMethod):
```
Comment (Ying1123) [inline python/sglang/srt/layers/linear.py]: same here. avoid using "if...else"

**Comment by jinyangyuan-nvidia** [comment]
> Is there any performance data to show the improvement?

The perf data have been updated in the description.

File: python/sglang/srt/layers/linear.py (hunk 1 above)
Comment (jinyangyuan-nvidia) [reply python/sglang/srt/layers/linear.py]: Thanks for the suggestion. Because modifying the first GEMM in the shared experts to used the CUTLASS backend only benefits small batch sizes, these modifications have been reverted.

File: python/sglang/srt/models/deepseek_v2.py [hunk 3]
```diff
@@ -298,8 +301,13 @@ def forward(self, hidden_states, gemm_output_zero_allocator: BumpAllocator = Non
             and self.weight.shape[0] == 256
             and _device_sm >= 90
         ):
-            # router gemm output float32
-            logits = dsv3_router_gemm(hidden_states, self.weight)
+            if _is_fp4_quantization_enabled():
+                # router gemm output float32
+                logits = dsv3_router_gemm(
+                    hidden_states, self.weight, out_dtype=torch.float32
```
Comment (ispobock) [inline python/sglang/srt/models/deepseek_v2.py]: Why don't use FP32 output for all cases?

File: python/sglang/srt/models/deepseek_v2.py [hunk 4]
```diff
@@ -364,14 +372,17 @@ def __init__(
             prefix=add_prefix("experts", prefix),
         )
 
+        correction_bias = self.gate.e_score_correction_bias
+        if _is_fp4_quantization_enabled():
+            correction_bias = correction_bias.to(torch.bfloat16)
```
Comment (ispobock) [inline python/sglang/srt/models/deepseek_v2.py:374]: The `e_score_correction_bias` is FP32 in the official weights, why do we need to convert it to BF16?
ref: https://github.com/sgl-project/sglang/pull/7825, https://github.com/sgl-project/sglang/pull/8197

File: python/sglang/srt/models/deepseek_v2.py (hunk 3 above)
Comment (Ying1123) [reply python/sglang/srt/models/deepseek_v2.py]: The original func call uses the default value for `out_dtype`, which is `torch.bfloat16`. Should it be explicitly specified?

Comment (Ying1123) [reply python/sglang/srt/models/deepseek_v2.py]: Actually, I think the kernel should not set a default value for dtype

Comment (jinyangyuan-nvidia) [reply python/sglang/srt/models/deepseek_v2.py]: > Why don't use FP32 output for all cases?

Just like @Ying1123's answer, it is because the default dtype is BF16 and FP32 output may not benefit cases other than NVFP4 that uses TRT-LLM Gen.

> Actually, I think the kernel should not set a default value for dtype

The code has been modified accordingly.

File: python/sglang/srt/models/deepseek_v2.py (hunk 4 above)
Comment (jinyangyuan-nvidia) [reply python/sglang/srt/models/deepseek_v2.py:374]: Because `trtllm_fp4_block_scale_moe` needs it to be BF16. If it is not converted here, the conversion will happen in every call of `trtllm_fp4_block_scale_moe`: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/moe/fused_moe_triton/layer.py#L1023.

Comment (ispobock) [reply python/sglang/srt/models/deepseek_v2.py:374]: So that means `trtllm_fp4_block_scale_moe` doesn't follow the original dtype in the weights?

File: python/sglang/srt/models/deepseek_v2.py [hunk 5]
```diff
@@ -298,8 +301,13 @@ def forward(self, hidden_states, gemm_output_zero_allocator: BumpAllocator = Non
             and self.weight.shape[0] == 256
             and _device_sm >= 90
         ):
-            # router gemm output float32
-            logits = dsv3_router_gemm(hidden_states, self.weight)
+            logits = dsv3_router_gemm(
+                hidden_states,
+                self.weight,
+                out_dtype=(
+                    torch.float32 if _is_fp4_quantization_enabled() else torch.bfloat16
```
Comment (ispobock) [inline python/sglang/srt/models/deepseek_v2.py]: I mean we can use `torch.float32` for all cases, not only for fp4 quantization. What do you think?

Comment (jinyangyuan-nvidia) [reply python/sglang/srt/models/deepseek_v2.py]: I am okay with it. Done.

File: python/sglang/srt/models/deepseek_v2.py (hunk 4 above)
Comment (jinyangyuan-nvidia) [reply python/sglang/srt/models/deepseek_v2.py:374]: https://github.com/flashinfer-ai/flashinfer/blob/main/csrc/trtllm_fused_moe_kernel_launcher.cu#L710

File: python/sglang/srt/models/deepseek_v2.py (hunk 3 above)
Comment (ispobock) [reply python/sglang/srt/models/deepseek_v2.py]: Actually the output of router gemm should be FP32 for better accuracy. @Fridge003 can we remove the BF16 default dtype?

Comment (Fridge003) [reply python/sglang/srt/models/deepseek_v2.py]: Sure~ Previously we need bfloat16 output since fp32 will cause some incompatibility bugs on MoE kernel. If these bugs can be solved then fixing it to fp32 seems OK to me

Comment (ispobock) [reply python/sglang/srt/models/deepseek_v2.py]: I think the bug should be fixed now. The topk kernel can support fp32 input: https://github.com/sgl-project/sglang/blob/bde73ee43f970126ecc641a65adb3cd50098793f/python/sglang/srt/layers/moe/topk.py#L669
You can double-check it and wait the ci passed.
